Setup.hs

```diff
@@ -70,27 +57,44 @@ rustBuildHook description localBuildInfo hooks flags = do
   -- run Rust build
   -- FIXME add cargo and rust to program db during configure step
   -- FIXME: add `--target $TARGET` flag to support cross-compiling to $TARGET
-  putStrLn "Call `cargo build --release` to build a dependency written in Rust"
-  rawSystemExit (fromFlag $ buildVerbosity flags) "cargo" ["build","--release"]
+  notice verbosity "Call `cargo build --release` to build a dependency written in Rust"
+  rawSystemExit (fromFlag $ buildVerbosity flags) "cargo"
+    [ "build"
+    , "--release"
+    , "--target-dir"
+    , rustTargetDir
+    ]
+
+  -- Install build results into cabal build directory
+  createDirectoryIfMissingVerbose verbosity True targetBuildDir
+  copyFileVerbose verbosity staticSource staticTarget
+  copyFileVerbose verbosity dynSource dynTarget
 
-  -- copy static lib
-  putStrLn $ "Copy libplonk_verify.a to " <> staticTarget
-  rawSystemExit (fromFlag $ buildVerbosity flags) "cp" [staticSource, staticTarget]
+  -- seems to be needed for local use of template haskell
+  copyFileVerbose verbosity dynSource dynTarget_
 
-  -- copy dyn lib
-  putStrLn $ "Copy libplonk_verify.dylib to " <> dynTarget
-  rawSystemExit (fromFlag $ buildVerbosity flags) "cp" [dynSource, dynTarget]
+  when (buildOS == OSX) $
+    rawSystemExit verbosity "install_name_tool"
+        [ "-id"
+        , "@rpath/" <> mkGenericSharedBundledLibName buildPlatform (compilerId c) targetLibname
+        , dynTarget
+        ]
 
-  putStrLn "rustc compilation succeeded"
+  info verbosity "rustc compilation succeeded"
   buildHook simpleUserHooks description localBuildInfo hooks flags
  where
+  verbosity = fromFlag $ buildVerbosity flags
   c = compiler localBuildInfo
+
   sourceLibname = "plonk_verify"
-  targetLibname = "Cplonk_verify"
-  sourceBuildDir = "target/release"
+  targetLibname = "C" <> sourceLibname
+  rustTargetDir = buildDir localBuildInfo <> "/rust-target"
+  sourceBuildDir = rustTargetDir <> "/release"
   targetBuildDir = buildDir localBuildInfo
   staticSource = sourceBuildDir <> "/" <> mkGenericStaticLibName sourceLibname
   staticTarget = targetBuildDir <> "/" <> mkGenericStaticLibName targetLibname
   dynSource = sourceBuildDir <> "/" <> "lib" <> sourceLibname <> "." <> dllExtension buildPlatform
   dynTarget = targetBuildDir <> "/" <> mkGenericSharedBundledLibName buildPlatform (compilerId c) targetLibname
 
+  dynTarget_ = targetBuildDir <> "/" <> mkGenericSharedLibName buildPlatform (compilerId c) targetLibname
```

Comment:
Is this actually needed?

Reply:
Unfortunately, it seems to be needed. It is used in line 74 to install the dylib under a different name. Without that line using template Haskell or the repl did not work for me. 

Cabal's `extra-bundled-libraries` feature appears to be broken in several ways (or at least very poorly document).